http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/HandlerRemovingChannelPool.java

```diff
@@ -55,37 +57,38 @@ public HandlerRemovingChannelPool(SdkChannelPool delegate) {
 
     @Override
     public Future<Void> release(Channel channel) {
-        removePerRequestHandlers(channel);
-        return delegate.release(channel);
+        return release(channel, new DefaultPromise<>(channel.eventLoop()));
     }
 
     @Override
     public Future<Void> release(Channel channel, Promise<Void> promise) {
-        removePerRequestHandlers(channel);
-        return delegate.release(channel, promise);
+        removePerRequestHandlers(channel).addListener(future -> delegate.release(channel, promise));
```

Comment:
Is there any risk? IIRC, we've seen issues where listener callback never got called

Reply:
I think you're referring to previous deadlock issues? I think in those cases it's because we were doing blocking calls like `await()` and `sync()`, which is why in this case used a listener instead of blocking on completion.